hazelcast/test/src/HazelcastTests8.cpp

```diff
@@ -1763,6 +1764,99 @@ namespace hazelcast {
+                    delete instance_;
+                }
+            };
+
+            HazelcastServer *connection_manager_translate::instance_ = nullptr;
+            const address connection_manager_translate::private_address{"localhost", 5701};
+            const address connection_manager_translate::public_address{"192.168.0.1", 5701};
+
+            TEST_F(connection_manager_translate, test_translate_is_used) {
+                // given
+                client_config config;
+                config.get_connection_strategy_config().get_retry_config().set_cluster_connect_timeout(
+                        std::chrono::seconds(1));
+                auto client = new_client(std::move(config)).get();
+                spi::ClientContext ctx(client);
+                connection::ClientConnectionManagerImpl connection_manager(ctx,
+                                                                           std::unique_ptr<connection::AddressProvider>(
+                                                                                   new test_address_provider{true}));
+
+                // throws exception because it can't connect to the cluster using translated public unreachable address
+                ASSERT_THROW(connection_manager.start(), exception::illegal_state);
+            }
+
+            TEST_F(connection_manager_translate, test_translate_is_used_when_member_has_public_client_address) {
+                // given
+                client_config config;
+                config.get_network_config().use_public_address(true);
+                config.get_connection_strategy_config().get_retry_config().set_cluster_connect_timeout(
+                        std::chrono::seconds(1));
+                auto client = new_client(std::move(config)).get();
+                spi::ClientContext ctx(client);
+
+                member m(public_address, ctx.get_hazelcast_client_implementation()->random_uuid(), false, {},
+                         {{endpoint_qualifier{1, "public"}, address{"127.0.0.1", 5701}}});
+
+                auto conn = ctx.get_connection_manager().get_or_connect(m);
+                ASSERT_TRUE(conn);
+            }
+
+            TEST_F(connection_manager_translate, test_translate_is_not_used_when_public_ip_is_disabled) {
+                // given
+                client_config config;
+                config.get_network_config().use_public_address(false);
+                config.get_connection_strategy_config().get_retry_config().set_cluster_connect_timeout(
+                        std::chrono::seconds(1));
+                auto client = new_client(std::move(config)).get();
+                spi::ClientContext ctx(client);
+
+                member m(public_address, ctx.get_hazelcast_client_implementation()->random_uuid(), false, {},
+                         {{endpoint_qualifier{1, "public"}, address{"127.0.0.1", 5701}}});
+
+                ASSERT_THROW(ctx.get_connection_manager().get_or_connect(m), std::exception);
```

Comment:
Is there a more specific exception type that is expected here? I think `std::exception` is too general.

Reply:
We receive boost::system_error with error_code `boost::asio::error::network_unreachable` on socket connect. This exception do not normally propagate to the user but we always wrap it for user API calls. But in this test scenario, we access an internal API and test it and hence we internally receive a boost exception. I changed to test for this specific error_code hoping that it will work for windows as well :)